Dynamically Linked Library in Presto CPP (#24330)
prestodb/presto: presto-native-execution/presto_cpp/main/dynamic_registry/README.md

# Dynamic Loading of Presto Cpp Extensions
This library adds the ability to load User Defined Functions (UDFs), connectors, or types without having to fork and build Prestissimo, through the use of shared libraries that a Prestissimo worker can access. These are to be loaded on launch of the Presto server. The Presto server searches for any .so or .dylib files and loads them using this library.
## Getting started

> 1. Create a cpp file for your dynamic library

Reviewer: Please indent the text and code examples under each numbered step. For example, "For dynamically loaded" should line up vertically with the beginning of "Create a cpp file" above it. 

In the PR, you can select the Files changed tab, then on the far right of this README.md file you can click the "...", select "View file", and see how the .md file will look when it is displayed. 

In the screenshot, you can see the text and code examples that should be indented, and that step 2 needs to be separated from step 1. 
<img width="939" alt="Screenshot 2025-02-11 at 10 18 15 AM" src="https://github.com/user-attachments/assets/627edabb-a309-4910-92da-e748469b9997" />

For dynamically loaded function registration, the format followed is mirrored of that of built-in function registration with some noted differences. Using [MyDynamicFunction.cpp](examples/MyDynamicFunction.cpp) as an example, the function uses the extern "C" keyword to protect against name mangling. A registry() function call is also necessary here.

> 2. Register functions dynamically by creating .dylib or .so shared libraries and dropping them in a plugin directory

Reviewer: ```suggestion
2. Register functions dynamically by creating .dylib or .so shared libraries and dropping them in a plugin directory

```

These shared libraries may be made using CMakeLists like the following:
```
add_library(name_of_dynamic_fn SHARED TestFunction.cpp)
target_link_libraries(name_of_dynamic_fn PRIVATE fmt::fmt Folly::folly gflags::gflags)
```

> 3. In the Prestissimo worker's config.properties file, set the plugin.dir property

Reviewer: ```suggestion
3. In the Prestissimo worker's config.properties file, set the plugin.dir property

```

Set the value of plugin.dir to the file path of the directory where the shared libraries are located.  
```
plugin.dir="User\Test\Path\plugin"
```

> 4. When the worker or the sidecar process starts, it will scan the plugin directory and attempt to dynamically load all shared libraries

Reviewer: ```suggestion
When the worker or the sidecar process starts, it scans the plugin directory and attempts to dynamically load all shared libraries.
```
I don't think of this as part of the steps to configure, it's what happens when things are run using the config in steps 1-3.
